### ISSUE TITLE
[CVE-2021-20066] : JSDom improperly allows the loading of local resources

### ISSUE THREAD

**Issue by guimard** [jsdom/jsdom#3124]
### Basic info:

- **Node.js version:** 15.x
- **jsdom version:** 16.4.0

### Minimal reproduction case

From https://fr.tenable.com/security/research/tra-2021-05 _([CVE-2021-20066](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-20066))_ :
> *Synopsis*
> 
> JSDom improperly allows the loading of local resources. Modern browser best practices dictate that the loading of local resources should be disallowed by default.
> 
> From documentation, JSDom does not, by default, load any subresources. Users must enable the loading of resources/subresources. For example, when creating a new JSDOM object, the resources item can be set to "usable" to allow the loading of external resources:
```js
const dom = new JSDOM(source, {url: "http://localhost:8080/", resources: "usable"});
```
> The issue here is that this setting also enables the loading of local resources. For example, the following code snippet verifies that JSDOM is attempting to access a local resource by using a non-existent file to throw an error:
```js
const jsdom = require("jsdom");
const { JSDOM } = jsdom;
source = '<iframe src="file:///does_not_exist" />';
const dom = new JSDOM(source, {url: "http://localhost:8080/", resources: "usable"});
console.log(dom.window.document.body.parentElement.outerHTML);
```
> Output when running the above:
```shell
$ node -v; npm -v; npm list | grep jsdom;
 v15.8.0
 7.5.1
 pocs@ /Users/user/research/browser/jsdom/pocs
 └── jsdom@16.4.0
 ~/r/b/j/pocs ❯❯❯ node local_resources.js
 <html><head></head><body><iframe src="file:///does_not_exist"></iframe></body></html>
 Error: Could not load iframe: "file:///does_not_exist"
 at onErrorWrapped (/Users/user/research/browser/jsdom/pocs/node_modules/jsdom/lib/jsdom/browser/resources/per-document-resource-loader.js:38:19)
 at Object.check (/Users/user/research/browser/jsdom/pocs/node_modules/jsdom/lib/jsdom/browser/resources/resource-queue.js:72:23)
 at /Users/user/research/browser/jsdom/pocs/node_modules/jsdom/lib/jsdom/browser/resources/resource-queue.js:124:14
 at processTicksAndRejections (node:internal/process/task_queues:94:5) [Error: ENOENT: no such file or directory, open '/does_not_exist']
{ errno: -2, code: 'ENOENT', syscall: 'open', path: '/does_not_exist' }
```

**Comment by domenic** [comment]
As discussed previously over email, this is equivalent to removing the brakes on your car (by explicitly opting in to loading all resources) and then complaining when the brake pedal does nothing. This is not a security issue, and the jsdom docs are very clear on the consequences of allowing resource loading.

This CVE is officially "disputed" and we hope that the folks filing this will withdraw it.

**Comment by domenic** [comment]
It is disabled by default.

**Comment by rugk** [comment]
GitHub now shows this old CVE [has been patched in v16.5.0](https://github.com/advisories/GHSA-f4c9-cqv8-9v98). In the [changelog here I see nothing about this](https://github.com/jsdom/jsdom/releases/tag/16.5.0) and I also don't know why it comes up now hmm…

**Comment by ChandraSannamuri** [comment]
This is still an open issue and showing in our twistlock reports

<img width="1167" alt="Image" src="https://github.com/user-attachments/assets/4d053a19-e5fd-4807-b5b6-bd09b8708eb0" />
